src/dotnet/CarbonAware.Tools.WattTimeClient/WattTimeClient.cs

```diff
@@ -0,0 +1,247 @@
+
+            using (var activity = ActivitySource.StartActivity())
+            {
+                activity?.AddTag(Constants.BalancingAuthorityAbbreviation, balancingAuthorityAbbreviation);
+
+                var result = await this.GetAsyncWithAuthRetries(url);
+
+                Log.LogDebug("For query {url}, received data {result}", url, result);
+
+                return JsonSerializer.Deserialize<List<GridEmissionDataPoint>>(result, options) ?? new List<GridEmissionDataPoint>();
+            }
+        }
+
+        /// <inheritdoc/>
+        public Task<IEnumerable<GridEmissionDataPoint>> GetDataAsync(BalancingAuthority balancingAuthority, string startTime, string endTime)
+        {
+            return this.GetDataAsync(balancingAuthority.Abbreviation, startTime, endTime);
+        }
+
+        /// <inheritdoc/>
+        public async Task<Forecast?> GetCurrentForecastAsync(string balancingAuthority){
+            var url = BuildUrlWithQueryString(
+                Paths.Forecast,
+                new Dictionary<string, string>() {
+                    { Constants.BalancingAuthorityAbbreviation, balancingAuthority }
+                });
+
+            Log.LogInformation("Requesting current forecast using url {url}", url);
+
+            using (var activity = ActivitySource.StartActivity())
+            {
+                activity?.AddTag(Constants.BalancingAuthorityAbbreviation, balancingAuthority);
+
+                var result = await this.GetAsyncWithAuthRetries(url);
+
+                Log.LogDebug("For query {url}, received data {result}", url, result);
+
+                return JsonSerializer.Deserialize<Forecast?>(result);
+            }
+        }
+
+        /// <inheritdoc/>
+        public Task<Forecast?> GetCurrentForecastAsync(BalancingAuthority balancingAuthority)
+        {
+            return this.GetCurrentForecastAsync(balancingAuthority.Abbreviation);
+        }
+
+        /// <inheritdoc/>
+        public async Task<IEnumerable<Forecast>> GetForecastByDateAsync(string balancingAuthority, string startTime, string endTime){
+            var url = BuildUrlWithQueryString(
+                Paths.Forecast,
+                new Dictionary<string, string>() {
+                    { Constants.BalancingAuthorityAbbreviation, balancingAuthority },
+                    { Constants.StartTime, startTime },
+                    { Constants.EndTime, endTime }
+                });
+
+            Log.LogInformation("Getting forecast by date using url {url}", url);
+
+            using (var activity = ActivitySource.StartActivity())
+            {
+                activity?.AddTag(Constants.BalancingAuthorityAbbreviation, balancingAuthority);
+
+                var result = await this.GetAsyncWithAuthRetries(url);
+
+                Log.LogDebug("For query {url}, received data {result}", url, result);
+
+                return JsonSerializer.Deserialize<List<Forecast>>(result, options) ?? new List<Forecast>();
```

Comment:
In cases where a list is expected, but we get no values, we return an empty list so that callers don't need to check to see if the list is null and has no values.  We guarantee they'll get a not null list, but it may be empty.  In cases where an object is returned, if there's no value, we return null.  For this reason, returns are consistent.